cryptarchia/cryptarchia.py

```diff
@@ -93,20 +143,57 @@ def block_position(self, block: BlockHeader) -> int:
                 return i
 
 
+@dataclass
+class LedgerState:
+    """
+    A snapshot of the ledger state up to some block
+    """
+
+    block: Id = None
+    nonce: Id = None
+    total_stake: int = None
+    commitments: set[Id] = field(default_factory=set)  # set of commitments
+    nullifiers: set[Id] = field(default_factory=set)  # set of nullified
+
+    def verify_committed(self, commitment: Id) -> bool:
+        return commitment in self.commitments
+
+    def verify_unspent(self, nullifier: Id) -> bool:
+        return nullifier not in self.nullifiers
+
+    def apply(self, block: BlockHeader):
+        assert block.parent == self.block
+        self.block = block.id()
+        self.nullifiers.add(block.leader_proof.nullifier)
+
+
 class Follower:
-    def __init__(self, genesis: BlockHeader, config: Config):
+    def __init__(self, genesis_state: LedgerState, config: Config):
         self.config = config
         self.forks = []
-        self.local_chain = Chain([genesis])
-
-    # We don't do any validation in the current version
-    def validate_header(block: BlockHeader) -> bool:
-        return True
+        self.local_chain = Chain([])
+        self.epoch = EpochState(
+            stake_distribution_snapshot=genesis_state,
+            nonce_snapshot=genesis_state,
+        )
+        self.ledger_state_snapshot = genesis_state
+        self.ledger_state = genesis_state
```

Comment:
Note, with this change there is no genesis-block, instead we have a genesis ledger-state, and the chain is empty.

Reply:
It's generally useful to have a genesis block that generates a ledger state, so that you can pass it around and blocks can always assume to have a parent

Reply:
I'm not convinced of that, it seems like the useful thing is knowing that every block has a "context" or "environment" (or state) that it is being evaluate against, therefore an original "context" makes sense to me.

Do you want to go with this approach until we see a case where a genesis block is useful?

Reply:
Yes, that's ok for me

Reply:
> I'm not convinced of that, it seems like the useful thing is knowing that every block has a "context" or "environment" (or state) that it is being evaluate against, therefore an original "context" makes sense to me.
> 

This is interesting, but, could you specify what would this `state` actually is compound of? I cannot get the full idea in my head 😅

Reply:
for now the genesis state holds
- initial coin distribution
- initial nonce for epoch randomness

Hopefully not much more than that in the final design.